feat: p2tr script path classification & validation

sangaman opened the pull request:
This adds logic to classify `p2tr` script path spend inputs according to the BIP 341 script validation rules. It also adds validation logic on the `p2tr` payment class to validate that a given witness stack matches the witness program commited to in the output script. It does this by validating the control block and merkle proof for inclusion of the selected tapscript in the taptree committed to in the witness program.

The new input classification and validation logic is tested against the following transaction which uses the script path to spend a taproot input found in the bitcoin signet chain.

https://explorer.bc-2.jp/tx/dda4cb9290415952d93f52518df978d45ae1710e60e711e6786926e99d305ef9?input:1

Lastly, this commit adds a sanity check to the `p2tr` extended validation to ensure there is not a mismatch between the address and output script in the event both are specified.

Ticket: BG-37366

sangaman commented on ts_src/taproot.ts:
```diff
@@ -266,3 +266,155 @@ export function getControlBlock(
+  }
+
+  // The last stack element is called the control block c, and must have length 33 + 32m,
+  // for a value of m that is an integer between 0 and 128[6], inclusive
```
Just realized this `[6]` got left in there, that's a footnote from the BIP 341 spec I should have removed.

OttoAllmendinger commented on ts_src/taproot.ts:
```diff
@@ -266,3 +266,155 @@ export function getControlBlock(
+  leafVersion: number;
+} {
+  // second to last element is the tapscript
+  const tapscript = witnessStack[witnessStack.length - 2];
```
assert length first

brandonblack commented on ts_src/payments/p2tr.ts:
```diff
@@ -186,13 +196,48 @@ export function p2tr(a: Payment, opts?: PaymentOpts): Payment {
+      if (witnessStack.length === 0) {
+        throw new TypeError('witness stack cannot be empty');
+      } else if (witnessStack.length === 1) {
+        // if there's only a single element it should be a key path spend schnorr signature
```
Should we allow an unsigned keypath here, ie. one item of length 0?

brandonblack commented on ts_src/payments/p2tr.ts:
```diff
@@ -117,16 +119,24 @@ export function p2tr(a: Payment, opts?: PaymentOpts): Payment {
     );
   });
 
+  const _witnessProgram = lazy.value(() => {
```
Should be `_scriptPubKey` not `_witnessProgram`

sangaman replied:
I'm not the best with the terms but I've been trying to follow the terminology from BIP 341:

> Let q be the 32-byte array containing the witness program (the second push in the scriptPubKey) 

As I understand this, the `scriptPubKey` would then refer to our `output`, but in this case I just want the 32 byte array which would be the `witnessProgram` aka the taproot pubkey. Let me know if that makes sense to you.

brandonblack replied:
I see. Hate, and was unaware of, this naming. It's a witness program, in that it refers to something that happens in the witness, but it's not a program at all. I _think_ it would be more consistent with other Payments to merge the logic, but will take a look.

brandonblack replied:
OK, I looked at the broader context, and I think that this value should be merged with `_taprootPubkey` but should only be calculated in this way if `a.output` exists (ie. not using `o.output` because that'd be circular).

sangaman replied:
Sounds good to me.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -266,3 +266,155 @@ export function getControlBlock(
+  tapscript: Buffer;
+  controlBlock: Buffer;
+  taptreeDepth: number;
+  internalPubkey: Buffer;
```
We might want to return the parity bit for the tweaked pubkey here too.

sangaman replied:
I'm still a bit confused, I don't think we have the parity bit in the control block to begin with, so we couldn't return it if we wanted. I thought we should assume the even Y coordinate point for this key.

sangaman replied:
This is done w/ my latest commit.

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -266,3 +266,155 @@ export function getControlBlock(
+ */
+export function isValidTapscript(
+  witnessStack: Buffer[],
+  witnessProgram: Buffer,
```
`scriptPubKey`

sangaman replied:
So the only issue with this is I'd run into a naming collision, since I call the value I get from tweaking the internal pubkey with the taptweak `taprootPubkey` within this method. So I'm inclined to leave the naming as is, so the value that comes from the witness is `witnessProgram` and the value that comes from tweaking the internal key is `taprootPubkey` and ultimately I check to make sure they're equal.

brandonblack replied:
> the value that comes from the witness is `witnessProgram`

But it comes from the `prevoutScript` not from the `witness`, so even though this is technically a correct name it's just rife for future confusion. `prevoutKey` `addressKey` ...

brandonblack replied:
Yeah, that might be best given how it's used in here.

Maybe you should call it like the BIP: `Let q be the 32-byte array containing the witness program (the second push in the scriptPubKey) which represents a public key according to BIP340.` (please don't)

brandonblack commented on ts_src/taproot.ts:
```diff
@@ -266,3 +266,155 @@ export function getControlBlock(
+    ]),
+  );
+
+  const taptreePath = [tapleafHash];
```
Why hold onto the path vs. just the current hash?

sangaman commented on ts_src/payments/p2tr.ts:
```diff
@@ -186,13 +197,47 @@ export function p2tr(a: Payment, opts?: PaymentOpts): Payment {
     return nameParts.join('-');
   });
+  lazy.prop(o, 'redeem', () => {
+    if (!a.redeems || a.redeemIndex === undefined) return;
```
Yeah we should but probably in a follow-up, it's not too hard to do but I'd want to add a few test cases as well which will take a bit of time.

sangaman commented:
I cleaned up the footnote and renamed `witnessProgram` in `isValidTapscript` to `expectedTaprootPubkey` and added `@param` comments detailing what it is, where it comes from, and how it's used.
